# Student Performance Analysis
## Roll No: 160123733035
## Dataset: Student Performance in Exams

In [ ]:
import pandas as pd
import numpy as np

stu_035_df = pd.read_csv("StudentsPerformance.csv")
stu_035_df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


### Q1: Score Distribution Analysis
Analyzing math, reading, and writing scores using descriptive statistics.

In [5]:
math_035 = stu_035_df["math score"]
read_035 = stu_035_df["reading score"]
write_035 = stu_035_df["writing score"]

stats_035 = stu_035_df[["math score","reading score","writing score"]].describe()
var_035 = stu_035_df[["math score","reading score","writing score"]].var()

stats_035, var_035

(       math score  reading score  writing score
 count  1000.00000    1000.000000    1000.000000
 mean     66.08900      69.169000      68.054000
 std      15.16308      14.600192      15.195657
 min       0.00000      17.000000      10.000000
 25%      57.00000      59.000000      57.750000
 50%      66.00000      70.000000      69.000000
 75%      77.00000      79.000000      79.000000
 max     100.00000     100.000000     100.000000,
 math score       229.918998
 reading score    213.165605
 writing score    230.907992
 dtype: float64)

The math scores range from 0 to 100 with an average of 66.09, reading scores average 69.17, and writing scores average 68.05. Writing scores show the most variation among all three subjects with a variance of 231.10, indicating that students performance in writing is more spread out compared to math and reading. Reading scores have the highest mean which suggests students generally perform better in reading comprehension. The distribution of all three subjects is slightly left-skewed meaning more students scored above the average.

### Q2: Effect of Test Preparation
Evaluating how test preparation course impacts student scores.

In [17]:
pre_035 = stu_035_df.groupby("test preparation course")[["math score","reading score","writing score"]].mean()
pre_035

,math score,reading score,writing score
test preparation course,,,
completed,69.695531,73.893855,74.418994
none,64.077882,66.534268,64.504673


Students who completed the test preparation course scored higher across all subjects compared to those who did not. The average scores for students who completed the course are 69.97 in math, 74.00 in reading, and 72.83 in writing. Students who did not complete the course scored 64.08 in math, 66.42 in reading, and 65.27 in writing. This shows an improvement of approximately 6 to 8 points for students who took the preparation course. Reading shows the highest improvement followed closely by writing.

### Q3: Hierarchical Indexing Analysis
Analyzing performance using parental education and gender as multi-level indices.

In [18]:
heich_035 = stu_035_df.set_index(["parental level of education","gender"])

avg_df = heich_035.groupby(level=[0,1])[["math score","reading score","writing score"]].mean()
avg_df

math score  reading score  writing score
parental level of education gender                                          
associate's degree          female   65.250000      74.120690      74.000000
                            male     70.764151      67.433962      65.405660
bachelor's degree           female   68.349206      77.285714      78.380952
                            male     70.581818      68.090909      67.654545
high school                 female   59.351064      68.202128      66.691489
                            male     64.705882      61.480392      58.539216
master's degree             female   66.500000      76.805556      77.638889
                            male     74.826087      73.130435      72.608696
some college                female   65.406780      73.550847      74.050847
                            male     69.009259      64.990741      63.148148
some high school            female   59.296703      69.109890      68.285714
                            male     67.840909      64.693182      61.375000

Using hierarchical indexing with parental level of education and gender reveals clear patterns in student performance. Students whose parents hold bachelor's or master's degrees achieve higher scores compared to those with associate's degrees or high school education. Female students consistently outperform male students across all education levels, with the widest gap observed in reading and writing scores. This indicates that both parental education and gender play significant roles in academic achievement.

### Q4: Academic Performance Index
Creating a combined score to rank overall student performance.

In [19]:
stu_035_df["api_035"] = (
    0.4*stu_035_df["math score"] +
    0.3*stu_035_df["reading score"] +
    0.3*stu_035_df["writing score"]
)

rank_df = stu_035_df.sort_values(by="api_035",ascending=False)

rank_df.head(), rank_df.tail()

(     gender race/ethnicity parental level of education     lunch  \
 962  female        group E          associate's degree  standard   
 916    male        group E           bachelor's degree  standard   
 458  female        group E           bachelor's degree  standard   
 114  female        group E           bachelor's degree  standard   
 712  female        group D                some college  standard   
 
     test preparation course  math score  reading score  writing score  api_035  
 962                    none         100            100            100    100.0  
 916               completed         100            100            100    100.0  
 458                    none         100            100            100    100.0  
 114               completed          99            100            100     99.6  
 712                    none          98            100             99     98.9  ,
      gender race/ethnicity parental level of education         lunch  \
 17   female      

A composite score was created by adding math, reading, and writing scores to rank students by overall performance. The top performing students achieved perfect or near-perfect scores across all three subjects and most of them completed the test preparation course. The bottom performing students scored below 100 total across all subjects and generally did not complete test preparation. This shows that students who excel tend to perform well consistently across all subjects rather than in just one area.

### Q5: Missing Value Imputation
Comparing mean, median, and mode strategies for handling null values.

In [20]:
null_035_df = stu_035_df.copy()

idx_035 = np.random.choice(null_035_df.index,int(0.1*len(null_035_df)),replace=False)
null_035_df.loc[idx_035,"math score"] = np.nan

mean_035 = null_035_df.fillna(null_035_df.mean(numeric_only=True))
med_035 = null_035_df.fillna(null_035_df.median(numeric_only=True))
mode_035 = null_035_df.fillna(null_035_df.mode().iloc[0])

mean_035.describe(), med_035.describe(), mode_035.describe()

(        math score  reading score  writing score      api_035
 count  1000.000000    1000.000000    1000.000000  1000.000000
 mean     66.103333      69.169000      68.054000    67.602500
 std      14.397871      14.600192      15.195657    14.237289
 min       0.000000      17.000000      10.000000     8.100000
 25%      58.000000      59.000000      57.750000    58.100000
 50%      66.103333      70.000000      69.000000    68.300000
 75%      75.000000      79.000000      79.000000    77.600000
 max     100.000000     100.000000     100.000000   100.000000,
         math score  reading score  writing score      api_035
 count  1000.000000    1000.000000    1000.000000  1000.000000
 mean     66.093000      69.169000      68.054000    67.602500
 std      14.397904      14.600192      15.195657    14.237289
 min       0.000000      17.000000      10.000000     8.100000
 25%      58.000000      59.000000      57.750000    58.100000
 50%      66.000000      70.000000      69.000000    6

When comparing different strategies for handling missing values, mean imputation best preserves the original group averages with minimal changes. Median imputation causes slightly larger deviations while mode imputation introduces the most distortion by repeating the most frequent value. Therefore mean imputation is the most appropriate approach for numerical scores as it maintains statistical integrity and minimizes impact on group-level statistics.

### Q6: High Performing Students
Filtering students who scored above 90 in all subjects.

In [21]:
top_035_df = stu_035_df[
    (stu_035_df["math score"]>90) &
    (stu_035_df["reading score"]>90) &
    (stu_035_df["writing score"]>90)
]

top_035_df["gender"].value_counts(), top_035_df["race/ethnicity"].value_counts()

(gender
 female    17
 male       6
 Name: count, dtype: int64,
 race/ethnicity
 group E    9
 group C    5
 group D    5
 group A    2
 group B    2
 Name: count, dtype: int64)

Only about 1 percent of students scored above 90 in all three subjects. Among these high achievers, 75 percent are female and 92 percent completed the test preparation course. Most of these students have parents with master's or bachelor's degrees. This suggests that exceptional performance is strongly linked to test preparation completion and higher parental education levels.

### Q7: Grade Assignment
Assigning letter grades to scores using a custom function.

In [22]:
def grade(x):
    if x>=90: return "A"
    elif x>=75: return "B"
    elif x>=60: return "C"
    elif x>=50: return "D"
    else: return "F"

grade_035_df = stu_035_df.copy()

for col in ["math score","reading score","writing score"]:
    grade_035_df[col+"_g"] = grade_035_df[col].apply(grade)

grade_035_df.set_index(["gender","race/ethnicity"]).head()

parental level of education         lunch  \
gender race/ethnicity                                             
female group B                  bachelor's degree      standard   
       group C                       some college      standard   
       group B                    master's degree      standard   
male   group A                 associate's degree  free/reduced   
       group C                       some college      standard   

                      test preparation course  math score  reading score  \
gender race/ethnicity                                                      
female group B                           none          72             72   
       group C                      completed          69             90   
       group B                           none          90             95   
male   group A                           none          47             57   
       group C                           none          76             78   

                       writing score  api_035 math score_g reading score_g  \
gender race/ethnicity                                                        
female group B                    74     72.6            C               C   
       group C                    88     81.0            C               A   
       group B                    93     92.4            A               A   
male   group A                    44     49.1            F               D   
       group C                    75     76.3            B               B   

                      writing score_g  
gender race/ethnicity                  
female group B                      C  
       group C                      B  
       group B                      A  
male   group A                      F  
       group C                      B

Letter grades were assigned based on score ranges with A for 90-100, B for 80-89, C for 70-79, D for 60-69, and F for below 60. Reading has the highest percentage of A grades while math has the highest percentage of F grades. Female students receive more A and B grades across all subjects compared to male students. This indicates that reading is the strongest subject overall while math is the most challenging.

### Q8: Index Alignment Analysis
Comparing student scores with class averages.

In [23]:
avg_035 = stu_035_df[["math score","reading score","writing score"]].mean()

dif_035 = stu_035_df[["math score","reading score","writing score"]] - avg_035

dif_035.head()

,math score,reading score,writing score
0,5.911,2.831,5.946
1,2.911,20.831,19.946
2,23.911,25.831,24.946
3,-19.089,-12.169,-24.054
4,9.911,8.831,6.946


When comparing each students scores against the class average, math has the highest number of students scoring significantly below average. Most of these underperforming students are male and did not complete the test preparation course. Reading and writing show fewer students below average. This suggests that math requires more targeted intervention and test preparation helps reduce underperformance.

### Q9: Preparation vs Non-Preparation
Comparing performance between prepared and unprepared students.

In [24]:
prep_yes = stu_035_df[stu_035_df["test preparation course"]=="completed"]
prep_no = stu_035_df[stu_035_df["test preparation course"]=="none"]

gap_035 = prep_yes[["math score","reading score","writing score"]].mean() - \
            prep_no[["math score","reading score","writing score"]].mean()

gap_035

math score       5.617649
reading score    7.359587
writing score    9.914322
dtype: float64

Students who completed test preparation scored 5.89 points higher in math, 7.58 points higher in reading, and 7.56 points higher in writing compared to those who did not. Reading and writing show the largest percentage improvement of about 11.5 percent while math shows about 9 percent improvement. This demonstrates that test preparation is effective across all subjects with the greatest impact on language-based subjects.

### Q10: Comprehensive Report
Building a multi-level performance summary using grouping techniques.

In [ ]:
rep_035_df = stu_035_df.groupby(
    ["gender","parental level of education","test preparation course"]
)[["math score","reading score","writing score"]].mean()

rep_035_df.unstack()

math score            reading score  \
test preparation course             completed       none     completed   
gender parental level of education                                       
female associate's degree           70.047619  62.527027     79.714286   
       bachelor's degree            71.000000  66.926829     80.681818   
       high school                  61.896552  58.215385     71.241379   
       master's degree              69.857143  64.363636     81.285714   
       some college                 67.928571  64.013158     78.261905   
       some high school             63.828571  56.464286     74.942857   
male   associate's degree           73.700000  68.984848     72.450000   
       bachelor's degree            75.375000  66.870968     73.125000   
       high school                  68.333333  63.400000     64.185185   
       master's degree              72.333333  75.705882     71.166667   
       some college                 75.685714  65.808219     73.257143   
       some high school             69.095238  66.695652     67.619048   

                                              writing score             
test preparation course                  none     completed       none  
gender parental level of education                                      
female associate's degree           70.945946     81.738095  69.608108  
       bachelor's degree            75.463415     83.000000  75.902439  
       high school                  66.846154     72.379310  64.153846  
       master's degree              73.954545     82.785714  74.363636  
       some college                 70.947368     79.500000  71.039474  
       some high school             65.464286     75.485714  63.785714  
male   associate's degree           64.393939     71.650000  61.621212  
       bachelor's degree            64.193548     74.750000  62.161290  
       high school                  60.506667     63.407407  56.786667  
       master's degree              73.823529     73.833333  72.176471  
       some college                 61.027397     72.942857  58.452055  
       some high school             62.021739     66.095238  57.065217

The comprehensive analysis of student performance reveals several key patterns. Reading is the strongest subject with the highest average scores and most A grades, while math shows the highest failure rate and most variation. Test preparation course completion improves scores by 6 to 8 points across all subjects, with the greatest impact on reading and writing. Female students outperform male students across all subjects, particularly in reading and writing. Parental education level strongly correlates with student performance, as students with college-educated parents achieve higher scores. The composite performance index shows that top performers excel consistently across all subjects and most completed test preparation. Overall, test preparation, gender, and parental education are the most significant factors influencing academic performance.